### DISCUSSION TITLE
How to run a workflow only once, even when triggered mutliple times

### DISCUSSION THREAD

**Discussion by a-recknagel** [community/community#26741] [Actions]
<p>I have a CI workflow that I want to run on PRs, but also when I’m just pushing to my repo. That way, I can notice issues sooner during development when I have not yet decided to turn it into a PR. The workflow also needs to run when a PR gets merged into master. But when I simply enter</p>
<pre><code>on: [push, pull_request]
</code></pre>
<p>it will run twice every time I update a PR that is on a branch that’s already part of the repo.</p>
<p>Either a config that fits my usecase or a way to tell a workflow to run only once given a commit hash would be nice.</p>

**Comment by oldskool** [comment]
<p>The way you have it setup it will always run on a push and when a PR is opened, yes. That would be expected when you configure it like that. If you only want to run it whenever you push a new branch (which is how I interpret your use case), you should listen just to the <a href="https://help.github.com/en/articles/events-that-trigger-workflows#create-event-create" rel="nofollow noopener">create event</a> instead, so you’ll end up with:</p>
<pre><code>on: [create, pull_request]
</code></pre>
<p>And it will only run when you first push a new branch and when you eventually open up a PR.</p>

**Comment by a-recknagel** [comment]
<p>That way it won’t be run when it gets merged into master, though. In other words, I want to run the workflow on every push for every branch within my repo, and also for PRs that come <em>from forked repos</em>. But reading the docs you linked for the <a href="https://help.github.com/en/articles/events-that-trigger-workflows#pull-request-events-for-forked-repositories" rel="nofollow noopener">pull request event</a>, it seems that there is no such configuration and i just might need to change the way I think about these triggers.</p>

**Comment by oldskool** [comment, marked as answer]
<p>In order to trigger things when they are merged into master, you can use the following workflow structure:</p>
<pre><code>on:
  pull_request:
    types: [closed]
    branches:
      - master

jobs:
  deploy:
    runs-on: ubuntu-latest

    steps:
      - name: checkout
        if: github.event.pull_request.merged
        uses: actions/checkout@v1
        with:
          ref: master
</code></pre>
<p>Note the “if” condition on the step. You will need to add that to every step to make sure it only run whenever the PR was actually merged (and not closed/discarded without merge). There is currently no way to enforce such a condition on all steps in a job yet, but this setup works just fine for my CD workflow (which I also only want to trigger when something is merged into master).</p>

**Comment by chrispat** [comment]
<p>We recently shipped support for if at the job level so this workflow should be easier now. </p>
<p><a href="https://help.github.com/en/articles/workflow-syntax-for-github-actions#jobsjob_idif" rel="nofollow noopener">https://help.github.com/en/articles/workflow-syntax-for-github-actions#jobsjob_idif</a></p>

**Comment by sypets** [comment]
<p>This has some more comments on running checks twice: <a href="https://github.community/t/duplicate-checks-on-push-and-pull-request-simultaneous-event/18012/4" class="inline-onebox">Duplicate checks on "push" and "pull_request" simultaneous event - #4 by yanjingzhu</a></p>
<p>Since most of the comments are older, I am not sure what is helpful, but it might be a good idea to link / merge these threads.</p>

**Comment by sypets** [comment]
<p>Personally, as a newbie I find the explanations confusing. The <a href="https://github.community/t/how-to-run-a-workflow-only-once-even-when-triggered-mutliple-times/16436/4">accepted solution above</a> only runs the CI on merge (which would make it not run at all when pushing to a feature branch in a PR, right?).</p>
<p>The <a href="https://github.community/t/how-to-run-a-workflow-only-once-even-when-triggered-mutliple-times/16436/5">newer comment</a> gives a helpful link, but there is no example.</p>

**Comment by tlylt** [comment]
<p>Example :<br>
<img src="https://user-images.githubusercontent.com/41845017/181099688-52a705a5-945f-4b80-9797-7aaee6fee443.png" alt="image" data-base62-sha1="oUgagJ4lo0NIQNqwlbchHxtA0Cg" width="356" height="132"></p>
